keps/sig-node/20190226-pod-overhead.md

```diff
@@ -93,157 +117,119 @@ so should not be impacted by pod Overhead.
-    // container resource usage. Users should leave this field unset.
+    // Overhead is the resource overhead incurred from the runtime.
     // +optional
     Overhead *ResourceRequirements
   }
 }
-```
 
-For scheduling, the pod resource requests are added to the container resource requests.
+For scheduling, the pod `Overhead` resource requests are added to the container resource requests.
 
 We don't currently enforce resource limits on the pod cgroup, but this becomes feasible once
 pod overhead is accountable. If the pod specifies a resource limit, and all containers in the
 pod specify a limit, then the sum of those limits becomes a pod-level limit, enforced through the
 pod cgroup.
 
-Users are not expected to manually set the pod resources; if a runtimeClass is being utilized,
-the manual value will be discarded. See RuntimeController for the proposal for setting these
-resources.
+Users are not expected to manually set `Overhead`; any prior values will being set will result in the workload
+being rejected. If runtimeClass is configured and selected in the PodSpec, `Overhead` will be set to the value
+defined in the corresponding runtimeClass. This is described in detail in
+[RuntimeClass admission controller](#runtimeclass-admission-controller).
+
+Being able to specify resource requirements for a workload at the pod level instead of container
+level has been discussed, but isn't proposed in this KEP.
+
+In the event that pod-level requirements are introduced, pod overhead should be kept separate. This
+simplifies several scenarios:
+ - overhead, once added to the spec, stays with the workload, even if runtimeClass is redefined
+ or removed.
+ - the pod spec can be referenced directly from scheduler, resourceQuote controller and kubelet,
+ instead of referencing a runtimeClass object which could have possibly been removed.
 
-Being able to specify resource requirements for a workload at the pod level instead of container level
-has been discussed, but isn't proposed in this KEP.
+#### Container Runtime Interface (CRI)
 
-In the event that pod-level requirements are introduced, pod overhead should be kept separate. This simplifies
-several scenarios:
- - overhead, once added to the spec, stays with the workload, even if runtimeClass is redefined or removed.
- - the pod spec can be referenced directly from scheduler, resourceQuote controller and kubelet, instead of referencing
- a runtimeClass object which could have possibly been removed.
+The pod cgroup is managed by the Kubelet, so passing the pod-level resource to the CRI implementation
+is not strictly necessary. However, some runtimes may wish to take advantage of this information, for
+instance for sizing the Kata Container VM.
+
+LinuxContainerResources is added to the LinuxPodSandboxConfig for both overhead and container
+totals, as optional fields:
+
+type LinuxPodSandboxConfig struct {
+	Overhead *LinuxContainerResources
```

Comment:
Is there a use case for providing overhead to the CRI? Otherwise it seems like just providing a pod total (sum of these 2 fields) would cover the use case of sizing the kata vm?

Reply:
@tallclair: the use case would being able to determine the non-overhead amount, actually.  In many (most?) cases, the overhead could account for hypervisor impact on the host, rather than guest sizing.

Reply:
I see. I think of the runtime as being the source-of-truth for the overhead, and the runtimeclass piece as just communicating that info to Kubernetes. I'm not strictly against it, but an alternative is to just have Kata subtract it's known overhead amount.

Also, isn't the guest OS overhead significant? Or is the overhead dominated by the hypervisor?

Reply:
> I see. I think of the runtime as being the source-of-truth for the overhead, and the runtimeclass piece as just communicating that info to Kubernetes. I'm not strictly against it, but an alternative is to just have Kata subtract it's known overhead amount.

It should, agreed, but system administrator could create their own runtimeClass definition, and the runtime wouldn't necessarily have visibility into what is being provided 'above.' Consider a scenario where the sys-admin defines two runtimeClasses, with differing overheads because one is for dedicated heavy IO, while another one is for compute workloads.   The OCI runtime wouldn't necessarily know in which context it is getting called.  This may not be a strong use case, but for a couple of bytes this helps leave the door open.

> Also, isn't the guest OS overhead significant? Or is the overhead dominated by the hypervisor?

It depends.  For a vCPU its negligible inside the guest, but more important on the host (ie, consider high bandwith IO, and the work a IO threads would carry out on the host).

Reply:
Makes sense. It still seems like the runtime needs some understanding of the overhead to know whether it should be allocated to the VM though. I'm happy to defer to you on this, but I'd like @yujuhong or @derekwaynecarr to sign off too.

Reply:
If there is no actionable intent right now to use the field, I prefer we not include it.  Right now, the use case seems like a soft hint.

Reply:
@derekwaynecarr : 

Well, I know how Kata would use it immediately, if available (which assumes CRI implementation passes this along).  I updated the actual text based on this before, as it wasn't very clear and lead to dialogue above. 

We (kata) would use the container-resources to help aid in creation of the sandbox (VM in kata scenario) up front.  Without this, the OCI runtime doesn't have a view of initial VM sizing, and would need to either: (1) hotplug CPUs as container creation requests are received after the initial sandbox creation request, or (2) add a lot of CPUs and constrain appropriately as we learn more.  (2) is terrible (inefficient and costly), and not what we do.  (1) works okay for more advanced hypervisors, but limits utility of `simpler` hypervisors (Firecracker being a good example of one which doesn't support hotplug).

The overhead, however, would need to be separate, as that is used for constraining on the host, but not for sizing the initial VM necessarily.

Reply:
@derekwaynecarr @tallclair see my comment above. I need to set the overhead as I implement RuntimeClass for Windows containers with Hyper-V isolation. There's no cgroup for me to inherit it from.

Reply:
Also, LinuxContainerResources may be too broad. How do you add/subtract non-absolute values such as cpu_period, cpu_shares, oom_score_adj, cpuset_cpus, cpuset_mems.

Some fields are also different from what's in [WindowsContainerResources](https://github.com/kubernetes/cri-api/blob/c67206fa03863cee5dcf257318fbb9acc3641850/pkg/apis/runtime/v1alpha2/api.proto#L653) - `cpu_count` isn't there on Linux, and `cpu_quota` is scaled differently from `cpu_maximum`.

I think it would be good to simplify this to have overhead just define 2 fields - `memory_size` and `cpu_quota` that will be added (or ignored if not implemented by the CRI)

Reply:
To put it another way - if Overhead isn't OS agnostic, then at least we would need

```
struct Overhead {
  LinuxContainerResources *LinuxContainerResources
  WindowsContainerResources *WindowsContainerResources
}
```

Reply:
Ah, I didn't take the windows sandbox into consideration here, this CRI API section is a bit Linux heavy.  Wouldn't the `WindowsContainerResources` only be useful in a Windows equivalent of the `LinuxPodSandboxConfig` ?

Reply:
Makes sense, thanks for weighing in Patrick. Here's my proposal: keep the current linux API as Eric has specified it, and add a parallel windows API:

```
message WindowsPodSandboxConfig {  // A new type, added to the PodSandboxConfig
  Overhead *WindowsContainerResources
  ContainerResources *WindowsContainerResources
}
```

As an aside, I don't love the field name "ContainerResources", but don't have a better suggestion...

Reply:
A separate `WindowsPodSandboxConfig` sounds good to me. 
And I'm terrible at naming :-|

Reply:
Sorry to chime in so late. I don't completely get why separate overhead and container resource fields are required. According the the thread below, they will be summed together for sizing the VM. Did I miss something?

Reply:
@yujuhong they aren't necessarily both used to size the VM.  The final text (not thread) should highlight this.  PTAL @ https://github.com/kubernetes/enhancements/pull/976